# Data Cleaning with Obeseity data 

Data cleaning is such an integral part of data analysis.Unlike on Kaggle,almost all data you see in the real world would be dirty and messy. Some even say data cleaning would take 80% of data analysis time.

The very fisrt step of any given data analysis project would be getting to know your data especially when you are dealing a messy one.

So, lets clean this messy data to start our analysis

## Import Libraries

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt
import seaborn as sns

## Read and Know Your Data

Take a look at our data as below, the data is quite obscure,it's hard to understand for a human,not mentioned for a computer.in this kind of situation,you have ways to get acquaintance with your data as follows:

1. Go to the data source page [WHO OBESITY DATA](https://apps.who.int/gho/data/node.main.A900A?lang=en)
2. If solution 1 doesn not work or hard to do, you can always go to ask data curator directly.

In [53]:
data=pd.read_csv('../data/data.csv')
data.head()

Unnamed: 0                                               2016  \
0          NaN  Prevalence of obesity among adults, BMI &Great...   
1          NaN                                         18+  years   
2      Country                                         Both sexes   
3  Afghanistan                                      5.5 [3.4-8.1]   
4      Albania                                   21.7 [17.0-26.7]   

                                              2016.1  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                               Male   
3                                      3.2 [1.3-6.4]   
4                                   21.6 [14.8-29.0]   

                                              2016.2  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                             Female   
3                                     7.6 [4.3-12.4]   
4                                   21.8 [15.3-28.9]   

                                                2015  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                         Both sexes   
3                                      5.2 [3.3-7.7]   
4                                   21.1 [16.6-26.0]   

                                              2015.1  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                               Male   
3                                      3.0 [1.3-6.0]   
4                                   20.9 [14.4-28.1]   

                                              2015.2  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                             Female   
3                                     7.3 [4.1-11.8]   
4                                   21.3 [15.1-28.1]   

                                                2014  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                         Both sexes   
3                                      4.9 [3.1-7.3]   
4                                   20.5 [16.2-25.1]   

                                              2014.1  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                               Male   
3                                      2.8 [1.2-5.6]   
4                                   20.2 [13.9-27.3]   

                                              2014.2  ...  \
0  Prevalence of obesity among adults, BMI &Great...  ...   
1                                         18+  years  ...   
2                                             Female  ...   
3                                     7.0 [4.0-11.3]  ...   
4                                   20.8 [14.9-27.4]  ...   

                                              1978.2  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                             Female   
3                                      0.9 [0.3-2.2]   
4                                     9.1 [4.6-15.5]   

                                                1977  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                                         Both sexes   
3                                      0.6 [0.2-1.2]   
4                                     6.8 [4.0-10.7]   

                                              1977.1  \
0  Prevalence of obesity among adults, BMI &Great...   
1                                         18+  years   
2                    

## Why and how the data is messy?

In most cases, data is collected by human or machines,a tiny glich would cause a long strip of bad data.if the data is collected by human, then it is a big chance that it would be messy. data can be dirty in many different ways,but mostly fall into those categories :

1. Missing data : like NAN
2. Validity of data : like 2016.1 / 2016.2 in the column
3. Outliers : like if a BMI entry is greater than 100
4. Consistency of data : the unit of every entry is not the same
5. Correctness of data: we are not gonna go through this ,but it is an important part of doing analysis in bussiness world, basically you need external data source or database to cross check the data in your hand because as we always say:

You dont know what you dont know

6. Data is in wide form not in long form : we are gonna go deeper about this one.

Are you ready, it's time to get our hands dirt

## long form VS wide form

The original data we have here is in wide form which means the form is very wide literally.

The .1 .2 in year number stand for gender, we gonna fix that later.

Wide data is not easy to analyze or stored effectively in computer, so we want to change it as soon as we can. go to read this tidy-data if you want to know more.



### Rename the columns appropriately and unpivot the data in the desirable format using pandas melt().

In [54]:
# Make a copy of the data
df=data.copy()
# Rename
df.rename(columns={'Unnamed: 0':'country'},inplace=True)
df=df.melt(id_vars=['country'],var_name='year')


In [55]:
df.head()

country  year                                              value
0          NaN  2016  Prevalence of obesity among adults, BMI &Great...
1          NaN  2016                                         18+  years
2      Country  2016                                         Both sexes
3  Afghanistan  2016                                      5.5 [3.4-8.1]
4      Albania  2016                                   21.7 [17.0-26.7]

## Correct the format

1. we will drop the first 3 row since its actually headers in the original forms.
2. correct year value
3. correct the gender value

In [56]:
# drop the first 3 columns
df=df.drop([0,1,2]).reset_index()

# spit the year columns into year and gender
df[['year','gender']]=df['year'].str.split('.',expand=True)\

# drop the index column
df=df.drop(columns=['index'])
df

country  year             value gender
0                             Afghanistan  2016     5.5 [3.4-8.1]   None
1                                 Albania  2016  21.7 [17.0-26.7]   None
2                                 Algeria  2016  27.4 [22.5-32.7]   None
3                                 Andorra  2016  25.6 [20.1-31.3]   None
4                                  Angola  2016    8.2 [5.1-12.2]   None
...                                   ...   ...               ...    ...
24940  Venezuela (Bolivarian Republic of)  1975   13.2 [7.1-20.7]      2
24941                            Viet Nam  1975     0.2 [0.1-0.5]      2
24942                               Yemen  1975     4.2 [1.5-8.9]      2
24943                              Zambia  1975     2.5 [1.0-5.1]      2
24944                            Zimbabwe  1975    6.7 [3.0-12.2]      2

[24945 rows x 4 columns]

In [57]:
# mapping the gender to the appropriate values
df['gender']=df['gender'].map({None:'both sex','1':'male','2':'female'})
df

country  year             value    gender
0                             Afghanistan  2016     5.5 [3.4-8.1]  both sex
1                                 Albania  2016  21.7 [17.0-26.7]  both sex
2                                 Algeria  2016  27.4 [22.5-32.7]  both sex
3                                 Andorra  2016  25.6 [20.1-31.3]  both sex
4                                  Angola  2016    8.2 [5.1-12.2]  both sex
...                                   ...   ...               ...       ...
24940  Venezuela (Bolivarian Republic of)  1975   13.2 [7.1-20.7]    female
24941                            Viet Nam  1975     0.2 [0.1-0.5]    female
24942                               Yemen  1975     4.2 [1.5-8.9]    female
24943                              Zambia  1975     2.5 [1.0-5.1]    female
24944                            Zimbabwe  1975    6.7 [3.0-12.2]    female

[24945 rows x 4 columns]

## correct the BMI value columns

From the webpage of WHO we can know that the values in [] are actually estimation intervel，so we need to seperate them into 3 columns

you can use str.matach() or str.findall() with regular expression to extract float number in this field,but we are gonna use str.split()

In [58]:
# Renaming the value column to BMI
df.rename(columns={'value':'BMI'},inplace=True)
df1=df.copy()

# Splitting the BMI value into BMI,upper and lower estimates
df['BMI']=df1['BMI'].str.split('[',expand=True)[0]
df['BMI_lower_esti']=df1['BMI'].str.split('[',expand=True)[1].str.split('-',expand=True)[0]
df['BMI_upper_esti']=df1['BMI'].str.split('[',expand=True)[1].str.split('-',expand=True)[1].str.split(']',expand=True)[0]
df=df[['country','year','gender','BMI','BMI_upper_esti','BMI_lower_esti']] #reset the column sequence
df.head()

country  year    gender    BMI BMI_upper_esti BMI_lower_esti
0  Afghanistan  2016  both sex   5.5             8.1            3.4
1      Albania  2016  both sex  21.7            26.7           17.0
2      Algeria  2016  both sex  27.4            32.7           22.5
3      Andorra  2016  both sex  25.6            31.3           20.1
4       Angola  2016  both sex   8.2            12.2            5.1

## Check validity of all columns or fields
We now have a pretty clean data compared to the one we just got. but our job is still not done yet. we need to go through every columns or fields to make sure the data is reletively correct.

**Country columns**

### What we know：

There is a country named country which need to be fixed

There are Nones in country column which need to be fixed

We have

### What we do：

We gonna drop those entries.

In [59]:
#locate Na
df.isna().sum()

country           250
year                0
gender              0
BMI                 0
BMI_upper_esti    879
BMI_lower_esti    879
dtype: int64

In [60]:
# dropping the rows wherever the values is 'Country'
df1=df.dropna(subset=['country'])
df1=df1.drop(df[df.country=='Country'].index)

In [61]:
df1.head()

country  year    gender    BMI BMI_upper_esti BMI_lower_esti
0  Afghanistan  2016  both sex   5.5             8.1            3.4
1      Albania  2016  both sex  21.7            26.7           17.0
2      Algeria  2016  both sex  27.4            32.7           22.5
3      Andorra  2016  both sex  25.6            31.3           20.1
4       Angola  2016  both sex   8.2            12.2            5.1

## BMI \ BMI_upper_esti and BMI_lower_esti columns

### What we know：

1. 4 contries have no BMI data which are Monaca,Sudan,South Sudan and San Marino,hence they dont have estimations.
2. We have 191 countries that do have BMI data and each of them has 126 entries.
3. The descriptive statistics of BMI data seems OK, no outliers.

### What we do：

1. We gonna create a new dataframe without those 4 countries to analyze.
2. We gonna change the data type of BMI and estimations to float.

In [39]:
print(df1.isna().sum())
print('\n')
print(df1[df1.BMI_upper_esti.isna()|df1.BMI_lower_esti.isna()].BMI.value_counts())
print('\n')
print(df1[df1.BMI=='No data'].country.value_counts())
con=df1[df1.BMI=='No data'].country.value_counts().index


country             0
year                0
gender              0
BMI                 0
BMI_upper_esti    504
BMI_lower_esti    504
dtype: int64


No data    504
Name: BMI, dtype: int64


San Marino     126
South Sudan    126
Sudan          126
Monaco         126
Name: country, dtype: int64


from above we can find that thare are 4 countries have no data, so exclude them, we have df2 to investigare



In [40]:
df2=df1[~df1.country.isin(con)]
# df2 dataframe is the one excluded no data countries

In [41]:
print(df2.country.value_counts().index)
val=df2.country.value_counts().values
np.unique(val)

Index(['Sierra Leone', 'Nauru', 'South Africa',
       'Bolivia (Plurinational State of)', 'Mongolia', 'Viet Nam', 'Thailand',
       'Jordan', 'Suriname', 'Ethiopia',
       ...
       'Saint Lucia', 'Italy', 'Burkina Faso', 'Grenada', 'Vanuatu', 'Japan',
       'Timor-Leste', 'Bosnia and Herzegovina', 'Luxembourg', 'Sri Lanka'],
      dtype='object', length=191)


array([126])

In [42]:
#change datatype to float
df2[['BMI','BMI_upper_esti','BMI_lower_esti']]=df2.loc[:,['BMI','BMI_upper_esti','BMI_lower_esti']].astype('float')

df2.loc[:,['BMI','BMI_upper_esti','BMI_lower_esti']].describe()

/home/greyatom/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:2963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


BMI  BMI_upper_esti  BMI_lower_esti
count  24066.000000    24066.000000    24066.000000
mean      12.448932       16.232112        9.237243
std       10.407428       12.003078        8.854281
min        0.100000        0.200000        0.000000
25%        3.900000        6.300000        2.200000
50%       10.600000       14.800000        7.000000
75%       18.175000       23.000000       13.800000
max       63.300000       70.800000       55.600000

## Activity (15 minutes)

## Visualization and EDA
Before you doing any EDA, come up with some questions first. Question orientated is always a good way to explore a set of data, you could easily fall into rabbit holes you enconter along the process otherwise.

What question we could possibly answer through this data?